cuda/test/base/math.cu

```diff
@@ -0,0 +1,139 @@
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/base/math.hpp>
+
+
+#include <cmath>
+#include <complex>
+#include <memory>
+
+
+#include <gtest/gtest.h>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/executor.hpp>
+
+
+#include "cuda/base/math.hpp"
+
+
+namespace {
+
+
+template <typename T>
+__global__ void test_real_isfinite(bool *result)
+{
+    constexpr auto inf = INFINITY;
+    bool test_true{};
+    bool test_false{};
+
+    test_true =
+        gko::isfinite(T{0}) && gko::isfinite(-T{0}) && gko::isfinite(T{1});
+    test_false =
+        gko::isfinite(inf) || gko::isfinite(-inf) || gko::isfinite(NAN);
+    *result = test_true && !test_false;
+}
+
+
+template <typename T>
+__global__ void test_complex_isfinite(bool *result)
+{
+    using c_type = thrust::complex<T>;  // std::complex<T>;
+    constexpr T inf = INFINITY;
+    constexpr T quiet_nan = NAN;
+    bool test_true{};
+    bool test_false{};
+
+    test_true = gko::isfinite(c_type{T{0}, T{0}}) &&
+                gko::isfinite(c_type{-T{0}, -T{0}}) &&
+                gko::isfinite(c_type{T{1}, T{0}}) &&
+                gko::isfinite(c_type{T{0}, T{1}});
+    test_false =
+        gko::isfinite(c_type{inf, T{0}}) || gko::isfinite(c_type{-inf, T{0}}) ||
+        gko::isfinite(c_type{quiet_nan, T{0}}) ||
+        gko::isfinite(c_type{T{0}, inf}) || gko::isfinite(c_type{T{0}, -inf}) ||
+        gko::isfinite(c_type{T{0}, quiet_nan});
+    *result = test_true && !test_false;
+}
+
+
+class IsFinite : public ::testing::Test {
+protected:
+    IsFinite()
+        : ref(gko::ReferenceExecutor::create()),
+          cuda(gko::CudaExecutor::create(0, ref))
+    {}
+
+    template <typename T>
+    bool test_real_isfinite_kernel()
+    {
+        gko::Array<bool> result(cuda, 1);
+        test_real_isfinite<T><<<1, 1>>>(result.get_data());
+        result.set_executor(ref);
+        return *result.get_data();
+    }
+
+    template <typename T>
+    bool test_complex_isfinite_kernel()
+    {
+        gko::Array<bool> result(cuda, 1);
+        test_complex_isfinite<T><<<1, 1>>>(result.get_data());
+        result.set_executor(ref);
+        return *result.get_data();
+    }
+
+    std::shared_ptr<gko::ReferenceExecutor> ref;
+    std::shared_ptr<gko::CudaExecutor> cuda;
+};
+
+
+TEST_F(IsFinite, Float) { ASSERT_TRUE(test_real_isfinite_kernel<float>()); }
+
+
+TEST_F(IsFinite, Double) { ASSERT_TRUE(test_real_isfinite_kernel<double>()); }
+
+
+TEST_F(IsFinite, FloatComplex)
+{
+    ASSERT_TRUE(test_complex_isfinite_kernel<float>());
```

Comment:
It is, however, extracting the floating point type is a bit annoying, and I thought it would be in the same file either way, but I can do that (will do the same for the CPU test then).